File: implementations/go/harness/shared_infra.go
package main

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// checkSharedInfra — [6] shared-infra: outbox / task-queue patterns
//
// If outbox/task-queue-related code exists, it must be grouped into a
// dedicated directory (named exactly "outbox"/"task-queue"), and that
// directory must actually implement the pattern's core types.
// shared-modules.md does not pin this directory's location to directly under
// internal/ — it may live under any concern-specific subpackage (e.g.
// internal/infrastructure/outbox/), so this recursively walks all of
// internal/ to check.
//
// Whether "the outbox pattern is in use" is determined by whether there is an
// actual reference to outbox.Writer (internal/infrastructure/persistence/
// etc.) — the transactional-save point that domain-events.md mandates for
// Repository implementations (and Application EventHandlers) — rather than
// being tripped up by an unrelated file whose name happens to contain the
// string "outbox" (e.g. migrations/000X_add_outbox.sql). Before the 2026-07
// async migration, the gate was based on a reference to OutboxRelay in the
// command package, but since Command Handlers no longer reference any
// outbox-related type at all under the synchronous-drain-forbidden rule, that
// signal disappeared — the gate has instead moved to the Repository ↔
// outbox.Writer connection, which always remains present.
func checkSharedInfra(root string) RuleResult {
	result := RuleResult{Section: "shared-infra"}
	result.Findings = append(result.Findings, checkOutboxPattern(root)...)
	result.Findings = append(result.Findings, checkTaskQueuePattern(root)...)
	return result
}

func checkOutboxPattern(root string) []Finding {
	usesOutboxWriter := false
	var outboxDirs []string
	walkErr := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			if d.Name() == "outbox" {
				outboxDirs = append(outboxDirs, path)
			}
			return nil
		}
		if !strings.HasSuffix(path, ".go") || strings.HasSuffix(path, "_test.go") {
			return nil
		}
		content, readErr := os.ReadFile(path)
		if readErr != nil {
			return nil
		}
		if strings.Contains(string(content), "outbox.Writer") || strings.Contains(string(content), "outbox.NewWriter") {
			usesOutboxWriter = true
		}
		return nil
	})
	if walkErr != nil {
		return []Finding{failFinding(root, "directory walk failed: "+walkErr.Error())}
	}

	if !usesOutboxWriter {
		return []Finding{skipFinding("no outbox pattern")}
	}

	if len(outboxDirs) == 0 {
		return []Finding{failFinding("internal/**/outbox/", "references outbox.Writer but there is no dedicated outbox/ directory")}
	}

	// Relay (synchronous drain) is not used — Poller (Outbox → SQS publishing)
	// and Consumer (SQS → Handler execution) split that role between them
	// (domain-events.md).
	hasWriter, hasPoller, hasConsumer := false, false, false
	for _, dir := range outboxDirs {
		walkErr := filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
			if err != nil || d.IsDir() || !strings.HasSuffix(path, ".go") {
				return nil
			}
			content, readErr := os.ReadFile(path)
			if readErr != nil {
				return nil
			}
			src := string(content)
			if regexp.MustCompile(`(?m)^type\s+Writer\s+struct\b`).MatchString(src) {
				hasWriter = true
			}
			if regexp.MustCompile(`(?m)^type\s+Poller\s+struct\b`).MatchString(src) {
				hasPoller = true
			}
			if regexp.MustCompile(`(?m)^type\s+Consumer\s+struct\b`).MatchString(src) {
				hasConsumer = true
			}
			return nil
		})
		if walkErr != nil {
			return []Finding{failFinding(dir, "directory walk failed: "+walkErr.Error())}
		}
	}

	if hasWriter && hasPoller && hasConsumer {
		return []Finding{passFinding("internal/**/outbox/ (Writer/Poller/Consumer implementation confirmed)")}
	}
	var missing []string
	if !hasWriter {
		missing = append(missing, "Writer")
	}
	if !hasPoller {
		missing = append(missing, "Poller")
	}
	if !hasConsumer {
		missing = append(missing, "Consumer")
	}
	return []Finding{failFinding("internal/**/outbox/", "outbox/ directory exists, but could not find type declaration(s) for "+strings.Join(missing, ", "))}
}

func checkTaskQueuePattern(root string) []Finding {
	hasTaskFile := false
	hasTaskDir := false
	walkErr := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		name := d.Name()
		pathSlash := filepath.ToSlash(path)
		if d.IsDir() {
			if name == "task-queue" {
				hasTaskDir = true
			}
			return nil
		}
		if strings.Contains(name, "task_queue") && !strings.Contains(pathSlash, "/task-queue/") {
			hasTaskFile = true
		}
		return nil
	})
	if walkErr != nil {
		return []Finding{failFinding(root, "directory walk failed: "+walkErr.Error())}
	}

	if !hasTaskFile {
		return []Finding{skipFinding("no task-queue pattern")}
	}
	if hasTaskDir {
		return []Finding{passFinding("internal/**/task-queue/")}
	}
	return []Finding{failFinding("internal/**/task-queue/", "task files exist, but there is no dedicated task-queue/ directory")}
}

File: implementations/fastapi/harness/rules/shared_infra.py
"""[7] shared-infra: outbox·task_queue

The outbox trigger is determined by "is there code that actually references
OutboxWriter" — this avoids being fooled by an unrelated file whose name happens to
contain the string "outbox". Be careful: in a structure where every actual file lives
inside src/outbox/, a "find a file outside it" condition could produce a false SKIP of
"no outbox pattern."

The outbox's components are OutboxWriter (loading) + OutboxPoller (publishing) +
OutboxConsumer (receiving) — the trigger symbol used is OutboxWriter.

The task_queue directory name uses an underscore — a hyphen isn't a valid Python
package/module name (`from ..task-queue.x import Y` is a SyntaxError), so it uses an
underscore for the same reason as the event-placement rule's
`application/integration_event/`.
"""

from __future__ import annotations

import os

from .common import RuleResult, failed, norm, passed, read, skipped


def check(root: str, py_files: list[str]) -> RuleResult:
    result = RuleResult("shared-infra")
    src_dir = os.path.join(root, "src")

    uses_outbox_writer = any("OutboxWriter" in read(f) for f in py_files)
    # Catches both an actual use of TaskOutboxWriter (a correctly-placed implementation)
    # and a misplaced file whose basename merely contains "task_queue" but which lives
    # outside the designated directory (reflecting the outbox's past "find a file outside
    # it" bug, both are detected from the start this time) as triggers.
    uses_task_outbox_writer = any("TaskOutboxWriter" in read(f) for f in py_files)
    has_misplaced_task_file = any(
        "task_queue" in os.path.basename(f) and "/task_queue/" not in norm(f) for f in py_files
    )
    has_task_file = uses_task_outbox_writer or has_misplaced_task_file

    if uses_outbox_writer:
        outbox_dir = os.path.join(src_dir, "outbox") if os.path.isdir(src_dir) else None
        if not outbox_dir or not os.path.isdir(outbox_dir):
            result.add(failed("src/outbox/", "References OutboxWriter, but there's no src/outbox/ directory"))
        else:
            checks = {
                "outbox_writer.py": os.path.isfile(os.path.join(outbox_dir, "outbox_writer.py")),
                "outbox_poller.py": os.path.isfile(os.path.join(outbox_dir, "outbox_poller.py")),
                "outbox_consumer.py": os.path.isfile(os.path.join(outbox_dir, "outbox_consumer.py")),
            }
            if all(checks.values()):
                result.add(
                    passed("src/outbox/ (confirmed the OutboxWriter/OutboxPoller/OutboxConsumer implementation)")
                )
            else:
                missing = [name for name, ok in checks.items() if not ok]
                result.add(
                    failed(
                        "src/outbox/",
                        "The src/outbox/ directory exists, but cannot find " + ", ".join(missing),
                    )
                )
    else:
        result.add(skipped("No outbox pattern"))

    if has_task_file:
        task_dir = os.path.join(src_dir, "task_queue") if os.path.isdir(src_dir) else None
        if not task_dir or not os.path.isdir(task_dir):
            result.add(failed("src/task_queue/", "A task file exists, but there's no src/task_queue/"))
        elif uses_task_outbox_writer:
            checks = {
                "task_outbox_writer.py": os.path.isfile(os.path.join(task_dir, "task_outbox_writer.py")),
                "task_outbox_poller.py": os.path.isfile(os.path.join(task_dir, "task_outbox_poller.py")),
                "task_consumer.py": os.path.isfile(os.path.join(task_dir, "task_consumer.py")),
            }
            if all(checks.values()):
                result.add(
                    passed(
                        "src/task_queue/ (confirmed the TaskOutboxWriter/TaskOutboxPoller/TaskConsumer implementation)"
                    )
                )
            else:
                missing = [name for name, ok in checks.items() if not ok]
                result.add(
                    failed(
                        "src/task_queue/",
                        "The src/task_queue/ directory exists, but cannot find " + ", ".join(missing),
                    )
                )
        else:
            result.add(passed("src/task_queue/"))
    else:
        result.add(skipped("No task_queue pattern"))

    return result
